traffic_ops/traffic_ops_golang/user/user.go

```
@@ -508,3 +578,483 @@ func (user *TOUser) InsertQuery() string {
+		"username": validation.Validate(user.Username, validation.Required),
+		"tenantID": validation.Validate(user.TenantID, validation.Required),
+	}
+
+	// Password is not required for update
+	if user.LocalPassword != nil {
+		_, err := auth.IsGoodLoginPair(user.Username, *user.LocalPassword)
+		if err != nil {
+			return err
+		}
+	}
+
+	return util.JoinErrs(tovalidate.ToErrors(validateErrs))
+}
+
+func Create(w http.ResponseWriter, r *http.Request) {
+	var userV4 tc.UserV4
+	var err error
+	inf, userErr, sysErr, errCode := api.NewInfo(r, nil, nil)
+	if userErr != nil || sysErr != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, sysErr)
+		return
+	}
+	defer inf.Close()
+
+	tx := inf.Tx.Tx
+	if err := json.NewDecoder(r.Body).Decode(&userV4); err != nil {
+		api.HandleErr(w, r, tx, http.StatusBadRequest, err, nil)
+		return
+	}
+	if err := validateUserV4(userV4); err != nil {
+		api.HandleErr(w, r, tx, http.StatusBadRequest, err, nil)
+		return
+	}
+	if err := postValidateV40(userV4); err != nil {
+		api.HandleErr(w, r, tx, http.StatusBadRequest, err, nil)
+		return
+	}
+
+	toUser := TOUser{
+		APIInfoImpl: api.APIInfoImpl{ReqInfo: inf},
+	}
+	toUser.User = userV4.Downgrade()
+
+	authorized, err := toUser.IsTenantAuthorized(inf.User)
+	if err != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, errors.New("checking tenant authorized: "+err.Error()))
+		return
+	}
+	if !authorized {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusForbidden, errors.New("not authorized on this tenant"), nil)
+		return
+	}
+
+	// Convert password to SCRYPT
+	*userV4.LocalPassword, err = auth.DerivePassword(*userV4.LocalPassword)
+	if err != nil {
+		api.HandleErr(w, r, tx, http.StatusBadRequest, err, nil)
+		return
+	}
+
+	var resultRows *sqlx.Rows
+	_, ok, err := dbhelpers.GetRoleIDFromName(tx, userV4.Role)
+	if err != nil {
+		api.HandleErr(w, r, tx, http.StatusInternalServerError, nil, errors.New("no ID exists for the supplied role name"))
```

this shouldn't swallow the error